databuilder/query_model.py

```diff
@@ -1,452 +1,356 @@
 
-    def __post_init__(self):
-        if self.has_categories:
-            raise NotImplementedError("Categorised codelists are currently unsupported")
+    # Containment is a special case: its right-hand side must be something vector-like i.e.
+    # something containing multiple values. To build a column whose values are vectors,
+    # use the `CombineAsSet` aggregation.
+    class In(Column):
+        lhs: Column
+        rhs: Union[Column, tuple[Value], Codelist]
 
-    def __repr__(self):
-        if len(self.codes) > 5:
-            codes = self.codes[:5] + ("...",)
-        else:
-            codes = self.codes
-        return f"Codelist(system={self.system}, codes={codes})"
+    class NotIn(In):
+        ...
+
+
+class Categorise(Column):
+    categories: dict[Value, Column]
+    default: Value
+
+    def __hash__(self):
+        # `categories` is a dict and so not hashable by default, but we treat it as
+        # immutable once created so we're fine to make it hashable
+        return hash((tuple(self.categories.items()), self.default))
 
 
-class ValueFromFunction(Value):
-    def __init__(self, *args):
-        self.arguments = args
+# We don't currently support this in the DSL or the Query Engine but we include it for
+# completeness
+class Join(ManyRowsPerPatientTable):
+    lhs: Table
+    rhs: Table
 
-    def _get_referenced_nodes(self):
-        return tuple(arg for arg in self.arguments if isinstance(arg, QueryNode))
 
+# VALIDATION
+#
+# The main thing we need to validate here is the "domain constraint". Columns and tables
+# which are in one-row-per-patient form can be combined arbitrarily because we can JOIN
+# using the patient_id and be sure that we're not creating new rows. But for operations
+# involving many-rows-per-patient inputs we need to ensure that they are all drawn from
+# the same table. (We call this the "domain" for set theoretic reasons which the margin
+# of this comment are too small to contain.)
 
-class DateDifference(ValueFromFunction):
-    def __init__(self, start, end, units="years"):
-        super().__init__(start, end, units)
-        self.units = units
+PATIENT_DOMAIN = object()
 
 
-class DateAddition(ValueFromFunction):
-    pass
+class DomainMismatchError(Exception):
+    ...
+
+
+def validate_node(node):
+    # TODO: Runtime type validation (i.e. only acceptable types are passed in), possibly
+    # using something like pydantic.
+
+    # TODO: Validation of the "inner" types i.e. a column may contain date values and we
+    # want to make it an error if you try to compare these to an int. And there are
+    # several places (e.g. a Filter condition) where we require suppied columns to have
+    # boolean type. We'll need the DSL to get these types from the schema and pass them
+    # in somehow.
+
+    # The one exception to the "common domain" rule is the Join operation which takes
+    # tables from two different domains and produces a new domain
+    if not isinstance(node, Join):
```

Comment:
This exception here is yucky, but I think it's correct. Maybe it would be clearer to include all domain-generating nodes (just tables?) in the exclusion -- I guess the others aren't excluded now because they don't have children.

Reply:
To be honest, I slapped in `Join` as a bit of a last minute addition just to show how we could handle it so I didn't spend much time looking for an elegant solution here. I'm sure there is a way of doing it more neatly though, as you say. We should probably work out how to represent in with the DSL and handle it in the QE first though!